ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -532,14 +532,13 @@ func (r *RayClusterReconciler) reconcilePods(instance *rayv1alpha1.RayCluster) e
 		}
 
 		runningPods := corev1.PodList{}
-		for _, aPod := range workerPods.Items {
-			if (aPod.Status.Phase == corev1.PodRunning || aPod.Status.Phase == corev1.PodPending) && aPod.ObjectMeta.DeletionTimestamp == nil {
-				runningPods.Items = append(runningPods.Items, aPod)
+		for _, pod := range workerPods.Items {
+			// TODO (kevin85421): We also need to have a clear story of all the Pod status phases, especially for PodFailed.
+			if isPodRunningOrPendingAndNotDeleting(pod) {
+				runningPods.Items = append(runningPods.Items, pod)
 			}
 		}
-		r.updateLocalWorkersToDelete(&worker, runningPods.Items)
 		diff := workerReplicas - int32(len(runningPods.Items))
-
 		if PrioritizeWorkersToDelete {
```

Comment:
Just for my education, Is there any conditions that we need to set `PrioritizeWorkersToDelete` to false?

The reason I am asking this is because without `PrioritizeWorkersToDelete`, if  diff >= 0, the kuberay operator will not respect the spec field(here `WorkersToDelete`) set by autoscaler.

Reply:
Perhaps I'm missing something, but for now, It seems to me:

- If the autoscaler is enabled, it is necessary to set `PrioritizeWorkersToDelete` to `true` to ensure that the autoscaler functions properly.
- if autoscaler is disabled, then the user is the only one who can update the `WorkersToDelete` field. In this case, the user effectively acts as the autoscaler. 

So, I can not think a situation that we need to set `PrioritizeWorkersToDelete` to `false`.

Reply:
Consider a situation with `PrioritizeWorkersToDelete` set to `false`:

1. at T0, we have 8 running workers.
2. at T1, user decides to remove worker A. So, he set replica = replica -1, and add worker A to `WorkersToDelete` in the yaml file.
3. at the same time T1, worker B and worker C fails because of OOM(let's see the restart policy is never).
4. at time T2, kuberay operator tries to collect all running workers and compare with the expected replica. it will find that the num current running pods is 8-2 = 6, the expected replica is 8 - 1 = 7. So, kuberay operator will add a new workers and never delete worker A.

Reply:
> Is there any conditions that we need to set PrioritizeWorkersToDelete to false?

`PrioritizeWorkersToDelete` is a feature gate (#208), and its default value is set to true. You can refer to the lessons we learned from #973. If we have a feature that allows users to enable or disable it, the cost of addressing the issue after the v0.5.0 release will not be that expensive. When the feature is stable enough, we can remove it.

>